c++/src/kj/async-io-test.c++

```diff
@@ -681,6 +681,52 @@ kj::Promise<void> expectRead(kj::AsyncInputStream& in, kj::StringPtr expected) {
   }));
 }
 
+class MockAsyncInputStream: public AsyncInputStream {
+public:
+  MockAsyncInputStream(kj::ArrayPtr<const byte> bytes, size_t blockSize)
+      : bytes(bytes), blockSize(blockSize) {}
+
+  kj::Promise<size_t> tryRead(void* buffer, size_t minBytes, size_t maxBytes) override {
+    // Clamp max read to blockSize.
+    size_t n = kj::min(blockSize, maxBytes);
+
+    // Unless that's less than minBytes -- in which case, use minBytes.
+    n = kj::max(n, minBytes);
+
+    // But also don't read more data than we have.
+    n = kj::min(n, bytes.size());
+
+    memcpy(buffer, bytes.begin(), n);
+    bytes = bytes.slice(n, bytes.size());
+    return n;
+  }
+
+private:
+  kj::ArrayPtr<const byte> bytes;
+  size_t blockSize;
+};
+
+KJ_TEST("AsyncInputStream::readAllText() / readAllBytes(limit)") {
+  kj::EventLoop loop;
+  WaitScope ws(loop);
+
+  auto bigText = strArray(kj::repeat("foo bar baz"_kj, 12345), ",");
+  size_t blockSizes[] = { 1, 4, 256, 4096, 8192, bigText.size() };
+
+  for (size_t blockSize: blockSizes) {
+    {
+      KJ_CONTEXT(blockSize);
+      MockAsyncInputStream input(bigText.asBytes(), blockSize);
+      KJ_EXPECT(input.readAllText().wait(ws) == bigText);
```

Comment:
Not testing with a limit here?

Seems like we actually want a two-dimensional test matrix, one to set the stream's block size, and one to choose the read limit? And then test readAllText() and readAllBytes() for each?

Reply:
That makes sense.